File: java/di-example/src/main/java/di/container/ApplicationContext.java
package di.container;

import di.annotations.Autowired;
import di.annotations.Component;
import java.lang.reflect.Constructor;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class ApplicationContext {
    private final Map<Class<?>, Object> beans = new HashMap<>();

    public ApplicationContext(Set<Class<?>> componentClasses) {
        for (Class<?> clazz : componentClasses) {
            createBean(clazz);
        }
    }

    private Object createBean(Class<?> clazz) {
        if (beans.containsKey(clazz)) {
            return beans.get(clazz);
        }

        try {
            if (!clazz.isAnnotationPresent(Component.class)) {
                throw new RuntimeException(clazz.getName() + "is not a component");
            }

            Constructor<?>[] constructors = clazz.getDeclaredConstructors();

            Constructor<?> targetConstructor = null;

            for (Constructor<?> constructor : constructors) {
                if (constructor.isAnnotationPresent(Autowired.class)) {
                    targetConstructor = constructor;
                    break;
                }
            }

            if (targetConstructor == null) {
                targetConstructor = clazz.getDeclaredConstructor();
            }

            Class<?>[] parameterTypes = targetConstructor.getParameterTypes();

            Object[] dependencies = new Object[parameterTypes.length];

            for (int i = 0; i < parameterTypes.length; i++) {
                dependencies[i] = createBean(parameterTypes[i]);
            }

            Object instance = targetConstructor.newInstance(dependencies);

            beans.put(clazz, instance);

            return instance;

        } catch (Exception e) {
            throw new RuntimeException("Failed to create bean: " + clazz.getName(), e);
        }
    }

    public <T> T getBean(Class<T> clazz) {
        return clazz.cast(beans.get(clazz));
    }
}

File: php/resume-builder-vanilla/src/Core/Container/ApplicationContext.php
<?php

namespace ResumeBuilderVanilla\Core\Container;

use ReflectionClass;
use ReflectionException;
use RuntimeException;

#[\Attribute]
class Component {}

#[\Attribute]
class Autowired {}

class ApplicationContext
{
    private array $beans = [];

    public function __construct(array $componentClasses)
    {
        foreach ($componentClasses as $class) {
            $this->createBean($class);
        }
    }

    public function createBean(string $class)
    {
        if (isset($this->beans[$class])) {
            return $this->beans[$class];
        }
        try {
            $reflectionClass = new ReflectionClass($class);
            // print_r($reflectionClass);
            $componentAttrs = $reflectionClass->getAttributes(Component::class);

            if (empty($componentAttrs)) {
                throw new RuntimeException("$class is not a component");
            }

            $constructor = $reflectionClass->getConstructor();

            $dependencies = [];

            if ($constructor) {

                $autowiredAttrs = $constructor->getAttributes(Autowired::class);

                foreach ($constructor->getParameters() as $parameter) {
                    $type = $parameter->getType();

                    if (!$type) {
                        throw new RuntimeException("Cannot resolve dependency{$parameter->getName()} in $class");
                    }

                    $dependencyClass = $type->getName();

                    $dependencies[] = $this->createBean($dependencyClass);
                    // var_dump($dependencies);
                }
            }

            $instance = $reflectionClass->newInstanceArgs($dependencies);
            $this->beans[$class] = $instance;

            return $instance;
        } catch (ReflectionException $e) {
            throw new RuntimeException("Failed to create bean: $class", 0, $e);
        }
    }

    public function getBean(string $class)
    {
        return $this->beans[$class] ?? null;
    }
}
